sopel/web.py

```diff
@@ -237,3 +237,40 @@ def iri_to_uri(iri):
     urlencode = urllib.urlencode
 else:
     urlencode = urllib.parse.urlencode
+
+
+def trim_url(url):
+    # clean trailing sentence- or clause-ending punctuation
+    while url[-1] in '.,?!\'":;':
+        url = url[:-1]
+
+    # clean unmatched parentheses/braces/brackets
+    for (opener, closer) in [('(', ')'), ('[', ']'), ('{', '}'), ('<', '>')]:
+        if url[-1] == closer and url.count(opener) < url.count(closer):
+            url = url[:-1]
+
+    return url
+
+
+def search_urls(text, exclusion_char=None, clean=False):
+    re_url = r'((?:http|https|ftp)(?::\/\/\S+))'
+    if exclusion_char is not None:
+        re_url = r'((?<!%s)(?:http|https|ftp)(?::\/\/\S+))' % (exclusion_char)
+
+    r = re.compile(re_url, re.IGNORECASE | re.UNICODE)
```

Comment:
It shouldn't be necessary to recompile the regex every time this function is called. Runtime changes to the `exclusion_char` (via `.set`) are the only things that could invalidate it, and at the moment I don't think supporting that is worth the efficiency loss.

Even if it's only running every time a link is found (with the arbitrary protocol support I'd like to add eventually, at least every line containing `://`), it's still not good to recompile regexes all the time for no reason. It's a bad habit. :)

Reply:
Don't worry, internaly, `re.compile` use caching: it won't re-process the same pattern twice.